src/memoize.js

```diff
@@ -46,45 +56,36 @@ function shouldRun(invalidatorMap, state) {
  * @param  {ORM} orm - a redux-orm ORM instance
  * @return {Function} `func` memoized.
  */
-export function memoize(func, equalityCheck = eqCheck, orm) {
+export function memoize(func, equalityCheck = defaultEqualityCheck, orm) {
     let lastOrmState = null;
-    let lastResult = null;
     let lastArgs = null;
-    const modelNameToInvalidatorMap = {};
+    let lastResult = null;
+    let allAccessedModels = {};
 
     return (...args) => {
-        const [dbState, ...otherArgs] = args;
-
-        const dbIsEqual = lastOrmState === dbState ||
-                           !shouldRun(modelNameToInvalidatorMap, dbState);
-
-        const argsAreEqual = lastArgs && otherArgs.every(
-            (value, index) => equalityCheck(value, lastArgs[index])
-        );
+        const [nextOrmState, ...otherArgs] = args;
 
+        const dbIsEqual = lastOrmState &&
+            (lastOrmState === nextOrmState
+                || accessedModelInstancesAreEqual(allAccessedModels, lastOrmState, nextOrmState)
+            );
+        const argsAreEqual = lastArgs && compareArgs(lastArgs, otherArgs, equalityCheck);
         if (dbIsEqual && argsAreEqual) {
             return lastResult;
         }
 
-        const session = orm.session(dbState);
+        const session = orm.session(nextOrmState);
         const newArgs = [session, ...otherArgs];
         const result = func(...newArgs);
 
-        // If a selector has control flow branching, different
-        // input arguments might result in a different set of
-        // accessed models. On each run, we check if any new
-        // models are accessed and add their invalidator functions.
-        session.accessedModels.forEach((modelName) => {
-            if (!modelNameToInvalidatorMap.hasOwnProperty(modelName)) {
-                modelNameToInvalidatorMap[modelName] = nextState =>
-                    lastOrmState[modelName] !== nextState[modelName];
-            }
-        });
-
         lastResult = result;
-        lastOrmState = dbState;
+        lastOrmState = nextOrmState;
         lastArgs = otherArgs;
 
+        allAccessedModels = {
+            ...allAccessedModels,
+            ...session.accessedModels,
```

Comment:
Since our accessedModels are nested by table, doesn't this change overwrite previously accessed models for the same table?

Reply:
Yeah, this should be merged more deeply than this.